server/src/main/java/org/opensearch/repositories/blobstore/ChecksumBlobStoreFormat.java

```diff
@@ -167,6 +175,61 @@ public void write(final T obj, final BlobContainer blobContainer, final String n
         blobContainer.writeBlob(blobName, bytes.streamInput(), bytes.length(), false);
     }
 
+    /**
+     * Writes blob with resolving the blob name using {@link #blobName} method.
+     * Leverages the multipart upload if supported by the blobContainer.
+     *
+     * @param obj                 object to be serialized
+     * @param blobContainer       blob container
+     * @param name                blob name
+     * @param compressor          whether to use compression
+     * @param listener            listener to listen to write result
+     */
+    public void writeAsync(
+        final T obj,
+        final BlobContainer blobContainer,
+        final String name,
+        final Compressor compressor,
+        ActionListener<Void> listener
+    ) throws IOException {
+        if (blobContainer instanceof AsyncMultiStreamBlobContainer == false) {
+            write(obj, blobContainer, name, compressor);
+            listener.onResponse(null);
+            return;
+        }
+        final String blobName = blobName(name);
+        final BytesReference bytes = serialize(obj, blobName, compressor);
+        final String resourceDescription = "ChecksumBlobStoreFormat.writeAsync(blob=\"" + blobName + "\")";
+        try (IndexInput input = new ByteArrayIndexInput(resourceDescription, BytesReference.toBytes(bytes))) {
+            long expectedChecksum;
+            try {
+                expectedChecksum = checksumOfChecksum(input.clone(), 8);
+            } catch (Exception e) {
+                throw new ChecksumCombinationException(
+                    "Potentially corrupted file: Checksum combination failed while combining stored checksum "
+                        + "and calculated checksum of stored checksum",
+                    resourceDescription,
+                    e
+                );
+            }
+
+            try (
+                RemoteTransferContainer remoteTransferContainer = new RemoteTransferContainer(
+                    blobName,
+                    blobName,
+                    bytes.length(),
+                    true,
+                    WritePriority.HIGH,
+                    (size, position) -> new OffsetRangeIndexInputStream(input, size, position),
+                    expectedChecksum,
+                    true
+                )
+            ) {
+                ((AsyncMultiStreamBlobContainer) blobContainer).asyncBlobUpload(remoteTransferContainer.createWriteContext(), listener);
```

Comment:
which threadpool would this use?

Reply:
For dedicated cluster, this threadpool should be sufficient. But on non dedicated cluster, translog and segment upload could interfere with cluster state upload. We would need dedicated threadpool for remote cluster state uploads.

Reply:
created an isseu to address this https://github.com/opensearch-project/OpenSearch/issues/9816